Playwright: deal with flakiness in the `Plans` page when using mobile viewport.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR addresses some flakiness in the `Plans` page on Calypso that is observed when using the mobile viewport, which is a fallout from #62158.

Key changes:
1. at the `ElementHelper` level, include a wait for the `networkidle` event to fire.
2. additional handling in the `PlansPage.clickTab` level to wait for asynchronous loading to complete.

Details:

Change 1 is meant to guard against any potential future regressions by adding a common baseline case when the viewport size is `mobile`. In this size, the NavTabs are hidden behind a pseudo-select dropdown, and this dropdown can be a) dismissed automatically by the page loading if clicked too soon, and b) clicks against the dropdown can be swallowed by the page if the page is not fully loaded.

Change 2 is meant to place additional waiting mechanism specific to the `PlansPage`, as various contents of the page are loaded asynchronously. Waiting for a `networkidle` event to fire is not sufficient here as the event fires despite there continuing to be placeholder elements on page.

#### Testing instructions

Ensure the following:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E
  - [x] i18n E2E


Related to #62158

worldomonation committed:
Add waits before clicking on the navbar in mobile view:
- for PlansPage only, wait for the network request to `plans?` to complete.
- for all `clickTab` actions, wait until the `networkidle` event is fired for mobile.

worldomonation committed:
Add guards before clicking on either tab.

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -51,8 +58,8 @@ export class PlansPage {
 	async validateActivePlanInMyPlanTab( expectedPlan: Plan ): Promise< void > {
-		await this.waitUntilLoaded();
-		await this.page.waitForSelector( selectors.myPlanTitle( expectedPlan ) );
+		const expectedPlanLocator = this.page.locator( selectors.myPlanTitle( expectedPlan ) );
+		await expectedPlanLocator.waitFor();
```
This is not strictly related to this change, but I touched code around here and so decided to apply up to date approach of using Locators here.

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -81,7 +88,19 @@ export class PlansPage {
 	 */
 	async clickTab( targetTab: PlansPageTab ): Promise< void > {
-		await this.page.waitForLoadState( 'networkidle' );
+		if ( targetTab === 'My Plan' ) {
+			// Current view is Plans, going to My Plans.
+			// Wait for the plans overview to render fully.
+			const locator = this.page.locator( selectors.plansView );
+			await locator.waitFor();
+		}
+		if ( targetTab === 'Plans' ) {
+			// Current view is My Plans, going to Plans.
+			// Wait for the details of current plan to
+			// render fully.
+			const placeholderLocator = this.page.locator( `.my-plan-card${ selectors.placeholder }` );
+			await placeholderLocator.waitFor( { state: 'hidden' } );
+		}
```
This may be slightly confusing as this takes the form of pre-processing prior to the click action on the intended tab, instead of the easier to understand post-processing after the tab has been clicked.

Because the whole reason why this is necessary is to prevent the async loading elements from swallowing up the `clickTab` action, I chose to go with the pre-processing as it would guarantee that the page state prior to the click is fully loaded (incl. async elements) and is ready to accept actions.

dpasque replied:
I think this makes a lot of sense!

My only concern: is there a different way we can gauge the starting spot? The method is currently written to guarantee the ending tab, but there's not technically a guarantee that we're starting on the other tab (people may call this not knowing what the default tab is they will land on).

Maybe we can check the URL? That seems pretty reliable!

worldomonation replied:
If I turn your question on the head a bit; why do we need to be mindful of the starting page, if the end state is guaranteed? 

This solution _should not_ get hung up if the source and destination tabs are the same, as I do not rely on network requests but rather that placeholders are all loaded.

dpasque replied:
Ah, I didn't explain myself well, sorry! Those are really good points.

The one, specific concern I had was... If someone is on the 'My Plan' tab, and they call this function targeting the same 'My Plan' tab, won't it throw an error because...
```typescript
const locator = this.page.locator( selectors.plansView );
await locator.waitFor();
```
... won't match on anything? Or am I misunderstanding?

worldomonation replied:
Ah okay, I think I get what you mean.

I did try out duplicating a call like this:

```
			await plansPage.clickTab( 'My Plan' );
			await plansPage.clickTab( 'My Plan' );
```

and I was able to cause it to hang, which should be the case you're referring to here.

I think maybe I'll try reviewing how this is done. There must be a better way to do this that doesn't rely on knowing where the originating tab is.

worldomonation replied:
I've pushed a much better solution in 60756d6cb549d76ae29c1a1647b680b562f3120f. I'd like to hear your thoughts.

dpasque replied:
Ahh yes, that short circuit approach is nice! LGMT! 🚀

worldomonation commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -51,6 +51,7 @@ export async function waitForElementEnabled(
 export async function clickNavTab( page: Page, name: string ): Promise< void > {
 	// Mobile view - navtabs become a dropdown.
 	if ( envVariables.VIEWPORT_NAME === 'mobile' ) {
+		await page.waitForLoadState( 'networkidle' );
```
While we have not had issues with other NavTabs yet, I felt it would not hurt to add a `networkidle` here. 

I'm not certain of the runtime impact of this statement and so am open to removing this as well.

worldomonation commented:
Ready for review.

dpasque reviewed (commented):
@worldomonation Looks good, but I had one small concern re: determining starting states. Let me know what you think!

worldomonation committed:
Revise how the clickTab action is pre-processed:
- add short circuit operation to bypass scenario where destination and current tab are the same.
- rework variable names.

dpasque reviewed (approved):
🎉
